### PR TITLE
Calculate tax before applying coupon so that it uses correct tax values

### PR DESCRIPTION
**Pull request by vedanshujain**
### All Submissions:

- [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
- [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
- [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

When creating "New order" from admin menu, if a new coupon is applies, and "Recalculate" button has not clicked on before, then the discount and total will be slightly inaccurate. This commit adds "calculate_tax" call before calculating discount to remove the inaccuracies.

Also removed a call for "calculate_total" because its already being called by "apply_coupon" method.

Towards #23802

### How to test the changes in this Pull Request:

1.  (From 23802) Add following settings in a new install:

(General) Enable tax rates and calculations
(General) Enable the use of coupon codes
(Tax) Yes, I will enter prices inclusive of tax
(Tax > Standard rates) Rate % 24.0000 

2. Create a new percentage coupon with 10% discount
3. Create a product with price 39 
4. Create a new order in WooCommerce Admin 
5. Add product created in step 3
6. Add coupon created in step 2

You will see that calculated prices are correct.

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [x] Have you successfully run tests with your changes locally?

### CODE REVIEW

**Review by peterfabian — COMMENTED** [review]
The PR fixes the issue in my testing, so I'm pre-approving with a question mark noted below. I understand we might want to merge it now in the current status due to shortage of time before the release, so asking for second opinion from the team @woocommerce/proton .

Btw, this 'Add order' screen behaves in a strange way also without the coupon, e.g. if user creates an order, adds the product and then hits `Create` [order] button, no taxes are added. Those are added only once the `Recalculate` button gets clicked. Maybe there's a reason for that, and if there is, it could also be a reason why applying the coupon behaves as it does now, before this PR?

**Review by kloon — CHANGES_REQUESTED** [review]
@vedanshujain would you be able to add some unit tests to cover this?

File: includes/class-wc-ajax.php
Comment (kloon) [reply includes/class-wc-ajax.php]: > Btw, this 'Add order' screen behaves in a strange way also without the coupon, e.g. if user creates an order, adds the product and then hits Create [order] button, no taxes are added. Those are added only once the Recalculate button gets clicked.

I believe this is just for flexibility, manual order creation is just that, a manual process that is why we give the options to automate certain tasks with buttons like the Recalculate but leave things as entered by the user creating the order.

With that said I am wondering if this change is needed then? I assume when you apply the coupon and click the recalculate button the totals are correct?

**Comment by kloon** [comment]
Had a chat with @peterfabian on slack and we are a bit unsure about these changes. As stated manual orders should give you the freedom to manually manage it, be that adding/removing taxes etc, if we automate calculations it might cause issues in the flow of certain users relying on the manual actions.

We learned the hard way with stock adjustments and automating that on manual orders so trying to avoid another situation like that.

We will evaluate the data for the next release and then decide to either proceed with this or rather just introduce notices on certain manual order actions being performed to keep the user informed.

**Comment by vedanshujain** [comment]
Sorry for late reply on this.

> if we automate calculations it might cause issues in the flow of certain users relying on the manual actions.

So the bug is if you add a product, and then apply a coupon, you would end up in an incorrect total state. You won't be able to get out of this state no matter how many time you click Recalculate button. 

This patch adds taxes when the coupon is applied. I guess if we were to keep the process manual, then I should be changing the behavior of the Recalculate button, to make sure recalculations are correct.

**Comment by daschenbrener** [comment]
I'm not sure is this a valid solution that works as I'm trying this on my 3.8 version of WC and it doesn't appear to change anything in the checkout.,

**Comment by peterfabian** [comment]
Since this hasn't been touched in in 9 months, I'm wondering if we should close this one down? Or do you prefer to keep this open @vedanshujain ?

**Comment by vedanshujain** [comment]
Yeah I am closing it. Not sure if this is valid anymore.

**Comment by niallmc1995** [comment]
Hey I am having a similar issue,
I have a tax rate of 12%, 
I also have a coupon for 10% off,
When I use the coupon it is taking from the tax rate also.
Anyone know how to fix this?

primptax-before-code
primptax
primptax-aftercode

**Pull request closed without merging**
